pkg/controllers/management/clusterprovisioner/helm_cleaner.go

```diff
@@ -0,0 +1,296 @@
+
+			if !replaced {
+				logrus.Infof("[cleanupHelmReleases] release %v in namespace %v has no deprecated or removed APIs", lastRelease.Name, namespace)
+				continue
+			}
+
+			if err := updateRelease(lastRelease, modifiedManifest, actionConfig); err != nil {
+				logrus.Errorf("[cleanupHelmReleases] failed to update release %v in namespace %v, skipping...", lastRelease.Name, lastRelease.Namespace)
+			}
+		}
+	}
+
+	return nil
+}
+
+// ReplaceManifestData replaces the out-of-date APIs with their respective valid successors, or removes an API that
+// does not have a successor.
+// Logic extracted from https://github.com/stormqueen1990/helm-mapkubeapis/blob/0245b7a7837a36fd164d83e496c453811d62c083/pkg/common/common.go#L81-L142
+func ReplaceManifestData(mapMetadata *mapping.Metadata, manifest string, kubeVersion string) (bool, string, error) {
+	var replaced = false
+	for _, mappingData := range mapMetadata.Mappings {
+		deprecatedAPI := mappingData.DeprecatedAPI
+		supportedAPI := mappingData.NewAPI
+		var apiVersion string
+
+		if mappingData.DeprecatedInVersion != "" {
+			apiVersion = mappingData.DeprecatedInVersion
+		} else {
+			apiVersion = mappingData.RemovedInVersion
+		}
+
+		if !semver.IsValid(kubeVersion) {
+			return replaced, "", errors.Errorf("Invalid format for Kubernetes semantic version: %v", kubeVersion)
+		}
+
+		if !semver.IsValid(apiVersion) {
+			return replaced, "", errors.Errorf("Failed to get the deprecated or removed Kubernetes version for API: %s", strings.ReplaceAll(deprecatedAPI, "\n", " "))
+		}
+
+		if count := strings.Count(manifest, deprecatedAPI); count > 0 {
+			if semver.Compare(apiVersion, kubeVersion) > 0 {
+				logrus.Debugf("The following API does not require mapping as the "+
+					"API is not deprecated or removed in Kubernetes '%s':\n\"%s\"\n", apiVersion,
+					deprecatedAPI)
+			} else {
+				if supportedAPI == "" {
+					logrus.Debugf("Found %d instances of deprecated or removed Kubernetes API:\n\"%s\"\n", count, deprecatedAPI)
+
+					for repl := 0; repl < count; repl++ {
+						// find the position where the API header is
+						apiIndex := strings.Index(manifest, deprecatedAPI)
+
+						// find the next separator index
+						separatorIndex := strings.Index(manifest[apiIndex:], "---\n")
+
+						// find the previous separator index
+						previousSeparatorIndex := strings.LastIndex(manifest[:apiIndex], "---\n")
+
+						/*
+						 * if no previous separator index was found, it means the resource is at the beginning and not
+						 * prefixed by ---
+						 */
+						if previousSeparatorIndex == -1 {
+							if apiIndex == 0 {
+								previousSeparatorIndex = 0
+							} else {
+								previousSeparatorIndex = apiIndex - 1
+							}
+						}
+
+						if separatorIndex == -1 { // this means we reached the end of input
+							manifest = manifest[:previousSeparatorIndex]
+						} else {
+							manifest = manifest[:previousSeparatorIndex] + manifest[separatorIndex+apiIndex:]
+						}
+					}
```

Comment:
Can this be place into a function for readability. It should be clear that this is the code that is mutating the manifest. Also, are we parsing yaml manually here? If so we can probably use a yaml parsing library instead unless speed is a demonstrable issue.

Reply:
@rmweir this method is mostly copypasta from https://github.com/helm/helm-mapkubeapis/blob/0245b7a7837a36fd164d83e496c453811d62c083/pkg/common/common.go, I understand your concerns and I resonate with them but ideally I wouldn't wish this code to diverge too far from helm-mapkubeapis so in the future we can just import it as a library and remove this entirely.

The reason for not using a YAML library is because it changes the Helm manifest a lot (by removing comments and other bits).

Reply:
Extracted part of it into a function.

Reply:
Okay that makes sense. Can you put a comment somewhere that eventually it can be removed and what version that should be possible on?

Reply:
The comment is already in-place; as for version, I guess it depends when my helm-mapkubeapis PR gets merged 😬